File: 41_maximum-subarray/maximum-subarray.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   ultimate010
@Problem:  http://www.lintcode.com/problem/maximum-subarray
@Language: Python
@Datetime: 16-06-13 09:27
'''

class Solution:
    """
    @param nums: A list of integers
    @return: An integer denote the sum of maximum subarray
    """
    def maxSubArray(self, nums):
        # write your code here
        if len(nums) == 0:
            return 0
        sums = [nums[0] for i in range(len(nums))]  # first num
        for pos in range(1, len(nums)):
            sums[pos] = max(sums[pos - 1] + nums[pos], nums[pos])
        return max(sums)

File: 41_maximum-subarray/maximum-subarray.cpp
/*
@Copyright:LintCode
@Author:   ultimate010
@Problem:  http://www.lintcode.com/problem/maximum-subarray
@Language: C++
@Datetime: 16-07-04 15:48
*/

class Solution {
public:    
    /**
     * @param nums: A list of integers
     * @return: A integer indicate the sum of max subarray
     */
    int maxSubArray(vector<int> nums) {
        // write your code here
        if (nums.size() == 0)
            return 0;
        int curSum = 0;
        int ret = nums[0];
        for (auto n : nums) {
            curSum = max(n, curSum + n);
            ret = max(ret, curSum);
        }
        return ret;
    }
};
